### DISCUSSION TITLE
Why does the default preconditoning(Block Diagonal Preconditioner) PMat not diagonal？

### DISCUSSION THREAD

**Discussion by nanguaxiaofendui** [idaholab/moose#23689] [Q&A General] [answered]
Hello all,
May I ask why does the output preprocessor matrix (PMat) in my program not block diagonal? Attached below is my program and the PMat.

[GlobalParams]
  displacements = 'disp_x disp_y'
  temperature = temperature
  volumetric_locking_correction = true 
[]

[Mesh]
  type = GeneratedMesh
  dim = 2
  nx = 4
  ny = 4
  xmax = 0.06 # Length of test chamber
  ymax = 0.06 # Test chamber radius
[]

[Variables]
  [./disp_x]
  [../]
  [./disp_y]
  [../]
  [temperature]
    initial_condition = 300 # Start at room temperature
  []
[]

[Kernels]
  [./TensorMechanics]
    use_automatic_differentiation = true
  [../]
  [heat_conduction]
    type = ADHeatConduction
    variable = temperature
  []
  [heat_conduction_time_derivative]
    type = ADHeatConductionTimeDerivative
    variable = temperature
  []
[]

[BCs]
  [confinex]
    type = DirichletBC
    variable = disp_x
    value = 0
    boundary = 'left'
  []
  [confiney]
    type = DirichletBC
    variable = disp_y
    value = 0
    boundary = left
  []
  [inlet_temperature]
    type = DirichletBC
    variable = temperature
    boundary = left
    value = 500 # (K)
  []
  [outlet_temperature]
    type = DirichletBC
    variable = temperature
    boundary = right
    value = 300 # (K)
  []
[]

[Materials]
  [./elasticity_tensor]
    type = ADComputeIsotropicElasticityTensor  
    youngs_modulus = 300e9
    poissons_ratio = 0.222
  [../]
  [./strain]
    type = ADComputeSmallStrain     
    eigenstrain_names = eigenstrain
    # List of eigenstrains to be applied in this strain calculation
  [../]

  [./thermal_strain]
    type = ADComputeThermalExpansionEigenstrain
    stress_free_temperature =300
    thermal_expansion_coeff = 8e-6
    eigenstrain_name = eigenstrain
  [../]
  [./stress]
    type = ADComputeLinearElasticStress
  [../]

  [./heat]  
    type = ADHeatConductionMaterial
    specific_heat =900
    thermal_conductivity = 27
  [../]

   [density]
    type = ADGenericConstantMaterial
    prop_names = 'density'
    prop_values = 3900
  []
[]

[Postprocessors]
  [single_nodal_temperature]
    type = PointValue
    point = '0.01 0.01 0'
    variable = temperature
  []
[]

[Executioner]
   type = Transient
#   scheme = bdf2
   end_time= 30
   solve_type =PJFNK
   petsc_options_iname = '-pc_type -pc_hypre_type -ksp_gmres_restart'
   petsc_options_value = 'hypre boomeramg 101'
[]

[Outputs]
   exodus = true
   csv = true
     [./console]
    	type = Console
    	output_file=true
  [../]
[]
![1678453417140](https://user-images.githubusercontent.com/126685008/224323070-45ebba3d-d7c6-4be7-ae1a-135e7d6729ff.png)

**Comment by GiudGiud** [comment, marked as answer]
Hello

you have coupling between 3 variables so there are diagonal terms in the Jacobian

the default preconditioning will use the Jacobian for PJFNK

Guillaume
